File: RecordWorkouts/views/AlertBuilder.m
#import "AlertBuilder.h"
#import <HealthKit/HKTypeIdentifiers.h>

@implementation AlertBuilder {
    UIAlertController *alert;
}

+ (void) showErrorAlertOn: (UIViewController *)delegate title:(NSString *)title error: (NSError *)error {
    NSString *message = error.code == HKErrorAuthorizationDenied
    ? @"Please enable access in\nSettings -> Privacy -> Health"
    : error.userInfo[NSLocalizedDescriptionKey];
    AlertBuilder *alertBuilder = [[AlertBuilder alloc] init:title message:message];
    [alertBuilder addOKAction];
    [alertBuilder show:delegate];
}

- (id) init: (NSString *)title message: (NSString *)message {
    self = [super init];
    
    alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    return self;
}

- (void) show: (UIViewController *)target {
    [target presentViewController:alert animated:YES completion:nil];
}

- (void) addDefaultAction: (NSString *)title handler: (void (^)(UIAlertAction * action))handler {
    UIAlertAction* action = [UIAlertAction actionWithTitle:title
                                                     style:UIAlertActionStyleDefault
                                                   handler:handler];
    [alert addAction:action];
}

- (void) addOKAction {
    UIAlertAction *ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:ok];
}

- (void) addCancelAction {
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel"
                                                     style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
}

@end

File: workout-records/views/AlertBuilder.swift
import UIKit
import HealthKit

class AlertBuilder {
    private var alert: UIAlertController
    
    class func showErrorAlert(on delegate: UIViewController, title: String, error: Error) {
        var message = error.localizedDescription
        if (error as NSError?)?.code == HKError.Code.errorAuthorizationDenied.rawValue {
            message = "Please enable access in\nSettings -> Privacy -> Health"
        }
        let alertBuilder = AlertBuilder(title, message: message)
        alertBuilder.addOKAction()
        alertBuilder.show(delegate)
    }
    
    class func showOKAlert(on delegate: UIViewController, title: String, message: String) {
        let alertBuilder = AlertBuilder(title, message: message)
        alertBuilder.addOKAction()
        alertBuilder.show(delegate)
    }
    
    init(_ title: String, message: String) {
        alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
    }
    
    func show(_ target: UIViewController) {
        target.present(alert, animated: true)
    }
    
    func addDefaultAction(_ title: String, handler: @escaping (_ action: UIAlertAction) -> Void) {
        let action = UIAlertAction(title: title, style: .default, handler: handler)
        alert.addAction(action)
    }
    
    func addOKAction() {
        addDefaultAction("OK", handler: { _ in })
    }

    func addCancelAction(_ handler: ((UIAlertAction) -> Void)?) {
        let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: handler)
        alert.addAction(cancel)
    }
}
